src/queue.rs

```diff
@@ -0,0 +1,148 @@
+                let mut next_job = ActiveModel::from(job);
+                next_job.parent_id = Set(Some(*active_model.id.as_ref()));
+                let next_job = next_job.insert(&tx).await?;
+                active_model.result = Set(Some(JobResult::Child(next_job.id)));
+            }
+
+            Ok(None) => {
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Success);
+                active_model.result = Set(Some(JobResult::Complete));
+            }
+
+            Err(e) if e.is_retryable() => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                let reschedule =
+                    schedule_based_on_failure_count(*active_model.failure_count.as_ref());
+                active_model.status =
+                    Set(reschedule.map_or(JobStatus::Failed, |_| JobStatus::Pending));
+                active_model.scheduled_at = Set(reschedule);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+
+            Err(e) => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Failed);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+        }
+        Some(active_model.update(&tx).await?)
+    } else {
+        None
+    };
+    tx.commit().await?;
+    Ok(model)
+}
+
+fn spawn(join_set: &mut JoinSet<()>, db: &Db, job_state: &Arc<SharedJobState>) {
+    let db = db.clone();
+    let job_state = Arc::clone(job_state);
+    join_set.spawn(async move {
+        loop {
+            match one(&db, &job_state).await {
+                Err(e) => {
+                    eprintln!("job error {e}");
+                }
+
+                Ok(Some(_)) => {}
+
+                Ok(None) => {
+                    sleep(Duration::from_secs(fastrand::u64(QUEUE_CHECK_INTERVAL))).await;
+                }
+            }
+        }
+    });
+}
+
+pub async fn run(db: Db, config: ApiConfig) {
+    let mut join_set = JoinSet::new();
+    let job_state = Arc::new(SharedJobState::from(&config));
+    for _ in 0..QUEUE_WORKER_COUNT {
+        spawn(&mut join_set, &db, &job_state);
+    }
+
+    while join_set.join_next().await.is_some() {
+        eprintln!("Worker task shut down. Restarting.");
+        spawn(&mut join_set, &db, &job_state);
+    }
+}
+
+async fn next(tx: &DatabaseTransaction) -> Result<Option<Model>, DbErr> {
+    let select = match tx.get_database_backend() {
+        backend @ DatabaseBackend::Postgres => Statement::from_sql_and_values(
+            backend,
+            r#"SELECT * FROM queue
+               WHERE status = $1 AND (scheduled_at IS NULL OR scheduled_at < $2)
+               ORDER BY updated_at ASC
+               FOR UPDATE
+               SKIP LOCKED
+               LIMIT 1"#,
```

Comment:
Is there anything in place to stop two workers from dequeuing the same job? What we do in Janus is have a lease column in the jobs tables so we can indicate that somebody currently has a lease on a job and thus other workers looking for jobs shouldn't take that one. The lease expires after some configurable amount of time to account for workers crashing while holding a lease.

Reply:
my understanding was that `select … for update` within a transaction will row-lock until the transaction is rolled back or committed, and the `skip locked` means other workers won't pick up the task until unlocked but I'll confirm the atomicity of that process in the docs.

Reply:
this continues to be my read of the relevant sections of the docs ([1](https://www.postgresql.org/docs/current/sql-select.html#SQL-FOR-UPDATE-SHARE), [2](https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-ROWS)

Reply:
This is my understanding as well, Janus leans on `FOR UPDATE SKIP LOCKED` in a similar way.